API/common/service_impl/chatgpt.py

```diff
@@ -23,29 +24,76 @@
 openai.api_base = os.environ["OPENAI_API_URL"]
 
 
-@timing_decorator
-def gpt_generate(system_msg: str, user_msg: str, history_msg: List[Dict[str, str]]) -> Dict[str, Any]:
+@log_dependency_call("gpt generate")
+def gpt_generate(context, system_msg: str, user_msg: str, history_msg: List[Dict[str, str]]) -> Dict[str, Any]:
     # the param dict of the chatgpt service
     chatgpt_service_params = initialize_chatgpt_service_params(system_msg)
     all_user_msg = []
+    estimated_history_tokens = 0
     if history_msg:
+        estimated_history_tokens = num_tokens_from_messages(history_msg)
         chatgpt_service_params["messages"].extend(history_msg)
         all_user_msg = [msg["content"]
                         for msg in history_msg if msg["role"] == "user"]
     all_user_msg.append(user_msg)
     chatgpt_service_params["messages"].append(
         {"role": "user", "content": "\n".join(all_user_msg)})
+    
+    # get the estimated number of tokens
+    estimated_prompt_tokens = num_tokens_from_messages(chatgpt_service_params["messages"])
+    gpt_task_name = context.custom_context.gpt_task_name
+    del context.custom_context.gpt_task_name
```

Comment:
Just out of curiosity, may I ask why we need to `del` the attributes in context?